server/lib/nl/detection/palm_api.py

```diff
@@ -0,0 +1,336 @@
+{
+  "PLACES":  ["nevada", "california", "north dakota"],
+  "METRICS": ["prevalence of asthma", "gender inequality", "per capita"],
+  "COMPARE": "COMPARE_PLACES"
+}
+```
+
+Input Sentence: "Countries of the world where poverty has increased the least?"
+Output JSON:
+```
+{
+  "PLACES":  ["world"],
+  "METRICS": ["poverty"],
+  "GROWTH": "INCREASE",
+  "RANK": "LOW",
+  "SUB_PLACE_TYPE": "COUNTRY"
+}
+```
+
+Input Sentence: "how big are cities in california?"
+Output JSON:
+```
+{
+  "PLACES":  ["california"],
+  "METRICS": ["size"],
+  "SIZE": "BIG",
+  "SUB_PLACE_TYPE": "CITY"
+}
+```
+
+Input Sentence: "prevalence of asthma vs. temperature rise in US cities where median age is over 30 and income is under $100K."
+Output JSON:
+```
+{
+  "PLACES": ["US"],
+  "METRICS": ["prevalence of asthma", "temperature rise"],
+  "SUB_PLACE_TYPE": "CITY",
+  "COMPARE": "COMPARE_METRICS",
+  "COMPARISON_FILTER": [
+    {
+      "COMPARISON_METRIC": "median age",
+      "COMPARISON_OPERATOR": "GREATER_THAN",
+      "VALUE": 30
+    },
+    {
+      "COMPARISON_METRIC": "income",
+      "COMPARISON_OPERATOR": "LESSER_THAN",
+      "VALUE": 100000
+    }
+  ]
+}
+```
+
+Input Sentence: "Blood pressure prevalence in US cities with highest income levels"
+Output JSON:
+```
+{
+  "PLACES": ["US"],
+  "METRICS": ["blood pressure prevalence"],
+  "SUB_PLACE_TYPE": "CITY",
+  "RANK": "HIGH",
+  "RANKING_FILTER": [
+    {
+      "RANKING_METRIC": "income levels",
+      "RANKING_OPERATOR": "IS_HIGHEST"
+    }
+  ]
+}
+```
+
+Input Sentence: "What are some of the biggest floods in the US?"
+Output JSON:
+```
+{
+  "PLACES": ["US"],
+  "RANK": "HIGH",
+  "DISASTER_EVENT": "FLOOD"
+}
+```
+
+Do not make up new JSON properties outside of the JSON Schema.
+""")
+
+_SUFFIX = '\n\nIn your response, include just the JSON adhering to the above schema. Do not add JSON keys outside the schema.  Also, explain why you set specific enum values.'
+
+# TODO: Consider tweaking this.
+_TEMPERATURE = 1.0
+
+_CANDIDATE_COUNT = 1
+
+_REQ_DATA = {
+    'prompt': {
+        'context': _PROMPT_CONTEXT_ORIG,
+        'messages': {},
+        'examples': [],
+    },
+    'temperature': _TEMPERATURE,
+    'candidateCount': _CANDIDATE_COUNT,
+}
+
+_SKIP_BEGIN_CHARS = ['`', '*']
+
+
+def call(query: str, history: str, ctr: counters.Counters) -> str:
+  logging.info(f'Doing PaLM API for {query}')
+  req_data = _REQ_DATA.copy()
+
+  if not history:
+    q = 'Convert this sentence to JSON: "' + query + '"'
+  else:
+    q = 'As a follow up to the last sentence, convert this sentence to JSON: "' + query + '"'
+  req_data['prompt']['messages']['content'] = q + _SUFFIX
+
+  for input, output in history:
+    req_data['prompt']['examples'].append({
+        'input': {
+            'content': input
+        },
+        'output': {
+            'content': json.dumps(output)
+        }
+    })
+
+  start_time = time.time()
+  req = json.dumps(req_data)
+  api_key = current_app.config['PALM_API_KEY']
```

Comment:
check api_key exists (or guard the caller with checking USE_PALM)?

Reply:
This is checked before calling llm_detector.  Added comment.